crates/extensions/c8y_mapper_ext/src/entity_cache.rs

```diff
@@ -0,0 +1,353 @@
+use std::collections::HashMap;
+use tedge_api::entity::EntityExternalId;
+use tedge_api::entity::EntityMetadata;
+use tedge_api::entity::EntityType;
+use tedge_api::entity::InsertOutcome;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::entity_store::EntityTwinMessage;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::pending_entity_store::PendingEntityData;
+use tedge_api::pending_entity_store::PendingEntityStore;
+use tedge_mqtt_ext::MqttMessage;
+use thiserror::Error;
+use tracing::debug;
+
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+#[error("Invalid external ID: {external_id} contains invalid character: {invalid_char}")]
+pub struct InvalidExternalIdError {
+    pub external_id: String,
+    pub invalid_char: char,
+}
+
+type ExternalIdMapperFn =
+    Box<dyn Fn(&EntityTopicId, &EntityExternalId) -> EntityExternalId + Send + Sync + 'static>;
+type ExternalIdValidatorFn =
+    Box<dyn Fn(&str) -> Result<EntityExternalId, InvalidExternalIdError> + Send + Sync + 'static>;
+
+/// Represents an error encountered while updating the store.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("The specified entity {0} does not exist in the store")]
+    UnknownEntity(String),
+
+    #[error(transparent)]
+    InvalidExternalId(#[from] InvalidExternalIdError),
+
+    #[error("Auto registration of the entity with topic id {0} failed as it does not match the default topic scheme: 'device/<device-id>/service/<service-id>'. Try explicit registration instead.")]
+    NonDefaultTopicScheme(EntityTopicId),
+}
+
+/// An in-memory cache of entity metadata with their external ids, indexed by their entity topic ids.
+/// The external id is the unique identifier of the entity twin in the connected cloud instance.
+/// This id is used
+/// Each entity in this cache is a mirror of the same entity in the entity store maintained by the agent,
+/// with the addition of the external id.
+///
+/// Every entity registered in this cache would have an external id which is either specified as the `@id`
+/// when the entity is registered or auto derived from the entity topic id.
+/// The user provided ids are validated using the `external_id_validator_fn` before they are added to the cache.
+/// When an `@id` is not provided, one is generated using the `external_id_mapper_fn`.
+///
+/// Any entity that is registered before its parents are cached in the `pending_entities` store,
+/// until those parents are registered as well.
+/// Once the parent is registered, the pending child entities are also registered along with it.
+pub(crate) struct EntityCache {
+    main_device_tid: EntityTopicId,
+    main_device_xid: EntityExternalId,
+    external_id_mapper_fn: ExternalIdMapperFn,
+    external_id_validator_fn: ExternalIdValidatorFn,
+
+    entities: HashMap<EntityTopicId, EntityMetadata>,
+    external_id_map: HashMap<EntityExternalId, EntityTopicId>,
```

Comment:
An option to address the concerns raised by https://github.com/thin-edge/thin-edge.io/pull/3230/files#r1907088899, is to:

- remove the `external_id` from the `EntityMetadata`
- and let the mapper manipulates `(EntityExternalId, EntityMetadata)` pairs
- these external ids being a mapper specific  concern.

```suggestion
    entities: HashMap<EntityTopicId, (EntityExternalId,EntityMetadata)>,
    external_id_map: HashMap<EntityExternalId, EntityTopicId>,
```

Not a simple change, though.

Reply:
Resolved by 5c917d1